Escape characters when generating file name for saving web page resources

szewai opened the pull request:
#### ef11a0fc671de16043e17c2c97af616d676cb809
<pre>
Escape characters when generating file name for saving web page resources
<a href="https://bugs.webkit.org/show_bug.cgi?id=267286">https://bugs.webkit.org/show_bug.cgi?id=267286</a>
<a href="https://rdar.apple.com/120501555">rdar://120501555</a>

Reviewed by Ryosuke Niwa and BJ Burg.

In existing implementation, file names generated for saved web page resources could contain &apos;%&apos; or other characters
that need to be encoded for URI. That means, when rewritting subresources URLs in the saved files, we have to encode
file name properly. To avoid this trouble, this patch generates file names that only contains unreserved characters for
URI (RFC 3986), which do not need to be percent-encoded in any case, and compatible with file system.

Updated API test: WebArchive.SaveResourcesValidFileName

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::isUnreservedURICharacter):
(WebCore::getFileNameFromURIComponent):
(WebCore::generateValidFileName):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: <a href="https://commits.webkit.org/272880@main">https://commits.webkit.org/272880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b161988b69d39e04cab924f7f681ecf0e62a4dd4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33487 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12257 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35406 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36110 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/30413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34497 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14611 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9416 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/29536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33962 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10334 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29880 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9033 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29872 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37432 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30402 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30218 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35268 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/9271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/7223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/33143 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11028 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7737 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/9856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10008 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -79,16 +80,44 @@ static const CFStringRef LegacyWebArchiveResourceTextEncodingNameKey = CFSTR("We
+    return isASCIIAlphanumeric(character) || character == '-' || character == '.' || character == '_' || character == '~';
+}
+
+static String encodeURIComponentForFileName(StringView input)
```
I think this name is confusing. Normally, `encodeURIComponent("a b c")` returns a%20b%20c, but this function does not actually encodeURIComponent, it transforms the filename to be filesystem-safe.

szewai replied:
The `encodeURIComponentForFileName` here means encoding a component extracted from URI for file name use, not encoding a component for URI use (as `encodeURIComponent` in JavaScript).
I see it could be a bit confusing; I am going to change it to `getFileNameFromURIComponent`.

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -79,16 +80,44 @@ static const CFStringRef LegacyWebArchiveResourceTextEncodingNameKey = CFSTR("We
+            result.append(character);
+            continue;
+        }
+        result.append('_');
```
I would use '-' as the default delimiter instead.

szewai committed:
Escape characters when generating file name for saving web page resources
https://bugs.webkit.org/show_bug.cgi?id=267286
rdar://120501555

Reviewed by Ryosuke Niwa and BJ Burg.

In existing implementation, file names generated for saved web page resources could contain '%' or other characters
that need to be encoded for URI. That means, when rewritting subresources URLs in the saved files, we have to encode
file name properly. To avoid this trouble, this patch generates file names that only contains unreserved characters for
URI (RFC 3986), which do not need to be percent-encoded in any case, and compatible with file system.

Updated API test: WebArchive.SaveResourcesValidFileName

* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::isUnreservedURICharacter):
(WebCore::getFileNameFromURIComponent):
(WebCore::generateValidFileName):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/CreateWebArchive.mm:

Canonical link: https://commits.webkit.org/272880@main
